pandas/tseries/index.py

```diff
@@ -1257,6 +1257,12 @@ def get_loc(self, key, method=None):
                                           'when key is a time object')
             return self.indexer_at_time(key)
 
+        # check if it's a Period and the frequencies are the same - otherwise a monthly period would match for
+        # a daily timestamp at the beginning of the month. NB: 'B' and 'D' therefore won't match
+        if isinstance(key, com.ABCPeriod) and key.freq == self.freq:
```

Comment:
I think this is ok *IF* the datetimeindex has a freq, but if its ``None`` (or a multiple) then what?

```
In [5]: '2013-03-31' in date_range('2013-01',periods=10,freq='M')[2:4:2]
Out[5]: True

In [10]: '2013-03-31' in date_range('2013-01',periods=10,freq='M')[2:4:2].union([Timestamp('2013-01-02')])
Out[10]: True
```